ddtrace/debugging/_origin/span.py

```diff
@@ -0,0 +1,268 @@
+        location = self.location
+        if root is None or span is None or root.get_tag("_dd.entry_location.file") is not None:
+            return self
+
+        # Add tags to the local root
+        for s in (root, span):
+            s.set_tag_str("_dd.ld.code_origin.type", "entry")
+
+            s.set_tag_str("_dd.ld.code_origin.frames.0.file", location.file)
+            s.set_tag_str("_dd.ld.code_origin.frames.0.line", str(location.start_line))
+            s.set_tag_str("_dd.ld.code_origin.frames.0.type", location.module)
+            s.set_tag_str("_dd.ld.code_origin.frames.0.method", location.name)
+
+        # TODO[gab]: This will be enabled as part of the live debugger/distributed debugging
+        # if ld_config.enabled:
+        #     # Create a snapshot
+        #     snapshot = Snapshot(
+        #         probe=location.probe,
+        #         frame=self.__frame__,
+        #         thread=current_thread(),
+        #         trace_context=root,
+        #     )
+
+        #     # Capture on entry
+        #     context = Debugger.get_collector().attach(snapshot)
+
+        #     # Correlate the snapshot with the span
+        #     root.set_tag_str("_dd.ld.code_origin.frames.0.snapshot_id", snapshot.uuid)
+        #     span.set_tag_str("_dd.ld.code_origin.frames.0.snapshot_id", snapshot.uuid)
+
+        #     self.set("context", context)
+        #     self.set("start_time", compat.monotonic_ns())
+
+        return self
+
+    def _close_context(self, retval=None, exc_info=(None, None, None)):
+        try:
+            context: SignalContext = self.get("context")
+        except KeyError:
+            # No snapshot was created
+            return
+
+        context.exit(retval, exc_info, compat.monotonic_ns() - self.get("start_time"))
+
+    def __return__(self, retval):
+        self._close_context(retval=retval)
+        return super().__return__(retval)
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        self._close_context(exc_info=(exc_type, exc_value, traceback))
+        super().__exit__(exc_type, exc_value, traceback)
+
+
+@dataclass
+class SpanCodeOriginProcessor(SpanProcessor):
+    _instance: t.Optional["SpanCodeOriginProcessor"] = None
+
+    def on_span_start(self, span: Span) -> None:
+        if span.span_type not in EXIT_SPAN_TYPES:
+            return
+
+        span.set_tag_str("_dd.ld.code_origin.type", "exit")
+
+        # Add call stack information to the exit span. Report only the part of
+        # the stack that belongs to user code.
+        seq = count(0)
+        for frame in frame_stack(sys._getframe(1), co_config.max_frame_depth):
+            code = frame.f_code
+            frame_origin = Path(code.co_filename)
+
+            if is_user_code(frame_origin):
+                print(frame_origin, "is user code")
+                n = next(seq)
+                span.set_tag_str(f"_dd.ld.code_origin.frames.{n}.file", str(frame_origin.resolve()))
```

Comment:
do we have method/type information here as well?

Reply:
no, unless we query the garbage collector, which is expensive :(